### ISSUE TITLE
Supporting C++20 in Envoy

### ISSUE THREAD

**Issue by martinezlucas98** [envoyproxy/envoy#15874]
*Supporting C++20 in Envoy*

*Description*:

Context:
The latest standard version ratified and published by ISO in December of 2020 as ISO/IEC 14882:2020 (informally known as C++20) —a very significant update over C++17—is widely (but not yet fully) supported by major compilers such as gcc-10 and clang-10.
C++20 is already being used as the standard version by some companies and this number will continue to increase as time passes, so an optional C++20 compatible version of Envoy would be needed.

What we're proposing:
Supporting C++20  by making the necessary changes, implementing the new features for better compatibility, and replacing deprecated features. We will be making these changes in a **modular way** as to not make the change so abrupt. We'll be starting with "unpopular" modules (those which are not widely used) and modules which require just a few changes to be fully compatible with C++.

- Ideas & suggestions on how to split up the PRs would be very helpful. Thanks.

**Comment by snowp** [comment]
>C++20 is already being used as the standard version by some companies and this number will continue to increase as time passes, so an optional C++20 compatible version of Envoy would be needed.

I'm not sure I follow why this means we need to be using C++20 in Envoy, surely most C++ compilers are backwards compatible? Can you speak to specific updates that you feel needs to be done in order to make Envoy compatible in these cases?

In any case I believe there have been concerns about supporting various platforms (e.g. for envoy-mobile), so not sure if we're ready to make the switch to C++20 yet. Maybe @jose @lizan @mattklein123 has thoughts here?

**Comment by moderation** [comment]
Recently the addition of a small C++17 feature caused build breakages and toil - https://github.com/envoyproxy/envoy/pull/14923 fixed by https://github.com/envoyproxy/envoy/pull/15093

**Comment by mattklein123** [comment]
Realistically I don't think we will be able to update to 20 any time soon, and will be gated on the mobile compilers. We can use this issue for tracking any work proposals.

**Comment by martinezlucas98** [comment]
> Yeah mobile compilers would be blocker, but would like to see if we can enable some C++20 features while not breaking C++17 compatibility (e.g. for some non-mobile used extensions).
@lizan 

Yes, that's exactly the idea: initial support without breaking C++17 compatibility while trying out C++20 features. This could also be useful/educational for people wanting to write new extensions in C++20.

Regarding Envoy-mobile, which directories (or sub-directories) would you consider mobile exclusive, to ensure we do not mess with them? Also, do you have any files/directories/sub-directories you think would be relevant for the issue we’re trying to modify, or any that you'd simply consider would be interesting to try out with C++20?

One more thing, do you happen to know if there's a way to make C++20 optional? Like a flag or something? In a way that if C++20 is not installed in the machine (or it’s not preferred by the user) then C++17 or something else will be used (note: we do not want to impose C++20).

Does this make sense? Any ideas and suggestions are really appreciated.

Thank you for your feedback! Apologies for the delayed response.

> This issue's title was also modified for better clarification.

**Comment by mattklein123** [comment]
The only way we would accept this right now is to allow C++20 to be opt-in for new extensions and no existing code, and even then we would need to discuss whether we want that extension to be ever used on Envoy Mobile.

IMHO it's not worth this right now, but if someone wants to do the build plumbing for this we can start there.

Stepping back, can you describe what features you want to use and how important they are?

**Comment by martinezlucas98** [comment]
@mattklein123 
> The only way we would accept this right now is to allow C++20 to be opt-in for new extensions and no existing code, and even then we would need to discuss whether we want that extension to be ever used on Envoy Mobile.
> 
The idea is exactly this, to use it for new extensions and use C++20 enabled libraries from this new extensions.

> IMHO it's not worth this right now, but if someone wants to do the build plumbing for this we can start there.
> 
That's exactly my motivation, to contribute the plumbing to enable experimentation with the new language features without affecting nor disrupting the core pieces.

> Stepping back, can you describe what features you want to use and how important they are?
>
We don't want to use any of the "super advance features", just start playing with the smaller convenient ones making way for the rest once support is ready and the timing is right.

Ideally, we'd contribute a small demo http extension to demonstrate how the potential use of C++ 20 on new extensions would work so that other envoy users can use this as a template.
Once that's done, we may contribute another more useful extension (not yet decided which) if you consider it reasonable when the time comes.

**Comment by mattklein123** [comment]
> We don't want to use any of the "super advance features", just start playing with the smaller convenient ones making way for the rest once support is ready and the timing is right.

I'm still not convinced there is value here. Any extension/code that uses c++20 is not going to be usable on Mobile. It would have to be a pretty compelling feature to make us ignore a portion of our users. Can you provide some more concrete detail on what features you want to use and why it's worth it to drop support for existing users for code that uses those features?

**Comment by martinezlucas98** [comment]
> I'm still not convinced there is value here. Any extension/code that uses c++20 is not going to be usable on Mobile. It would have to be a pretty compelling feature to make us ignore a portion of our users. Can you provide some more concrete detail on what features you want to use and why it's worth it to drop support for existing users for code that uses those features?
>
Hello @mattklein123 , it's just as you said before, my motivation is to contribute to the plumbing with a demo extension, so when the time for C++20 comes, there is this demo extension to take as an example maybe.
If by features you mean the C++20 features, there are some interesting ones like the `contains` member function for associative containers (as sets and maps); also, now we can bring enum's members into scope to improve readability.

Our goal is just a new and simple http extension demo using C++20, nothing too fancy.

I think this issue's title may be misleading, do you think I should change it to "C++20 demo extension"?

**Comment by mattklein123** [comment]
> I think this issue's title may be misleading, do you think I should change it to "C++20 demo extension"?

I guess? I'm still not really seeing the value here.
